Improvements to Frustum

mar-celine opened the pull request:
## Proposed changes

* Improve documentation in Frustum
* Extend map to allow for redistribution of gridpoints

A future PR will extend the map to allow the usage of equiangular maps in the future BBH domain.
Depends on #1421 

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

mar-celine commented:
Force push 2/16 14:13: clang-tidy and test_frustum fixes

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,179 @@ class er;
+ * \f]
+ *
+ *
+ * By default, the grid spacing between the two bases of the frustum is linear,
```
Do you mean 'uniform' instead of 'linear'?

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,179 @@ class er;
+ *
+ *
+ * By default, the grid spacing between the two bases of the frustum is linear,
+ * leading to grid distortion as the ratio between the sizes of the bases
```
What do you mean by 'grid distortion'?  I'm guessing that you mean that each element near the smaller base is stretched in the direction normal to the base and each element near the larger base is compressed in the direction normal to the base.  This should be made clearer, and there should be some statement saying that we desire each element to have roughly the same size in all 3 dimensions.

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,179 @@ class er;
+ * \f$\tilde{x}, \tilde{y}, \tilde{z}\f$ by \f$w\f$.
+ *
+ * We define an auxiliary variable `projective_zeta` \f$ = \mathrm{Z}\f$ such
+ * that the Cartesian coordinates are linear in \f$\mathrm{Z}\f$, and such that
```
linear -> uniform?

And what are the Cartesian coordinates? x with tildes or x without tildes?

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,179 @@ class er;
+ * The parameter \f$w_{\delta} = \f$ `projective_scale_factor` determines the
+ degree of projection
+ *of
+ * the \f$z_2\f$-base of the frustum into a fourth spatial dimension \f$w\f$.
```
add "as explained below" (because it's not clear at all what is meant or why you introduce a 4th spatial dimension).

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,179 @@ class er;
+ * to the frustum is an affine one. The user can choose to change the
+ * `projective_scale_factor`, increasing the foreshortening in the grid.
+ *
+ * For convenience, we provide the value of \f$w_{\delta}\f$ that corresponds to
```
"For convenience, we provide the value..." -> "Most often one will want to use the value..." 

("we provide" sounds like there is code for this choice)

markscheel replied:
I think it would be less confusing to move the discussion of this special value of w_delta elsewhere, like at the end.  Currently this discussion interrupts the flow of someone trying to figure out what the projective mapping is.

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,179 @@ class er;
+ * frustum, respectively. Other values for `projective_scale_factor` are
+ permitted if the user
+ * wants to use frustums for which the rectangular bases are not related by
+ * a scaling, for which there is not a unique `projective_scale_factor` value.
```
Should there be a constructor that computes this value of projective_scale_factor automatically?

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,179 @@ class er;
+ * a scaling, for which there is not a unique `projective_scale_factor` value.
+ *
+ * The gridpoint distribution is determined by using \f$\zeta\f$ to linearly
+ * interpolate in the homogeneous coordinates
```
"using \f$\zeta\f$ to linearly interpolate in the homogeneous coordinates" is unclear.  Do you mean that you choose grid points uniformly distributed in w?

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,179 @@ class er;
+ * recovering the Cartesian coordinates by dividing
+ * \f$\tilde{x}, \tilde{y}, \tilde{z}\f$ by \f$w\f$.
+ *
+ * We define an auxiliary variable `projective_zeta` \f$ = \mathrm{Z}\f$ such
```
Is this paragraph supposed to be explaining what "using zeta to linearly interpolate" means?  Or is this supposed to be another step in the calculation?

carmaza replied:
I think this should be a new paragraph regardless of its purpose (but making its purpose explicit, as in defining Z first, and explaining why it is useful)

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,179 @@ class er;
+ * by the relation:
+ * \f[ \vec{x}(\mathrm{Z}) = \frac{\tilde{\vec{x}}(\zeta)}{w(\zeta)}\f]
+ * Defining the auxiliary variables \f$w_{\pm} := w_{\delta}\pm 1\f$,
+ * \f$\mathrm{Z}(\zeta)\f$ is given by:
```
Another example of why I find this confusing.  When Z is discussed above, I'm unsuccessfully trying to figure out what Z is, and then later I see that Z is defined all the way down here.  Please define Z first.

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,179 @@ class er;
+ * \f$\mathrm{Z}\f$ reduces to \f$\zeta\f$ when \f$w_{\delta}\ = 1\f$.
+ * \f$\mathrm{Z}\f$ is then rationally interpolated by \f$\zeta\f$, as defined
+ * by the relation:
+ * \f[ \vec{x}(\mathrm{Z}) = \frac{\tilde{\vec{x}}(\zeta)}{w(\zeta)}\f]
```
why is xtilde a function of zeta?  it doesn't seem to be when it is defined above.

And why is w a function of zeta?  It wasn't a function of zeta above as far as I can tell.

It would help alot if when xtilde and w are first introduced, they were defined as functions of zeta.

markscheel commented on src/Domain/Creators/FrustalCloak.hpp:
```diff
@@ -0,0 +1,118 @@
+                 ProjectionFactor, LengthInnerCube, LengthOuterCube>;
+
+  static constexpr OptionString help{
+      "Creates an excised cube domain using frusta. The parameter\n"
```
First sentences should be "Creates a cubical domain with two equal-sized abutting excised cubes in the center. This is done by combining ten frusta".

Then the rest of the string doesn't need to mention abutting cubes or ten frusta.

markscheel commented on src/Domain/Creators/FrustalCloak.hpp:
```diff
@@ -0,0 +1,118 @@
+namespace creators {
+
+/// \ingroup DomainCreatorsGroup
+/// Create a 3D Domain consisting of frusta. Each frustum is made of two
```
Should be "Create a 3D cubical domain with two equal-sized abutting excised cubes in the center. This is done by combining ten frusta."

markscheel commented on src/Domain/Creators/FrustalCloak.hpp:
```diff
@@ -0,0 +1,118 @@
+  static constexpr OptionString help{
+      "Creates an excised cube domain using frusta. The parameter\n"
+      "`UseEquiangularMap` can be used to apply a tangent mapping to the xi\n"
+      "and eta logical coordinates, while the parameter `ProjectionFactor` \n"
```
coordinates -> coordinates of each frustum

markscheel commented on src/Domain/Creators/FrustalCloak.hpp:
```diff
@@ -0,0 +1,118 @@
+      "Creates an excised cube domain using frusta. The parameter\n"
+      "`UseEquiangularMap` can be used to apply a tangent mapping to the xi\n"
+      "and eta logical coordinates, while the parameter `ProjectionFactor` \n"
+      "can be used to apply a projective map to the zeta logical coordinate.\n"
```
coordinate -> coordinate of each frustum.

markscheel commented on src/Domain/Creators/FrustalCloak.hpp:
```diff
@@ -0,0 +1,118 @@
+      "Note that the shape of the excised region is not that of a cube, but\n"
+      "of two abutting cubes. This is because this Domain is primarily for\n"
+      "testing the frustal cloak in the BinaryCompactObject Domain."};
+
```
Say that the abutting excised cubes share a face in the x direction (or y or z, whichever it is).

markscheel reviewed (changes requested):
I find the documentation of the projective scale factor confusing because it is never clear whether a new paragraph is talking about a new step in the calculation, or whether that paragraph is trying to explain in more detail something that was said in a previous paragraph.  Similarly, various details are sprinkled about in seemingly random places, such as the definitions of the Delta and Sigma quantities which are defined quite early and then never used until the last two equations.

mar-celine commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,179 @@ class er;
+ * \f[\Delta^y_{\zeta} = \frac{1}{4}(y^{+\zeta}_{-\xi,-\eta} +
+ * y^{+\zeta}_{+\xi,+\eta} - y^{-\zeta}_{-\xi,-\eta} - y^{-\zeta}_{+\xi,+\eta})
+ * \f]
+ * \f[\Delta^y_{\eta}+ = \frac{1}{4}(y^{+\zeta}_{+\xi,+\eta} -
```
Stray `+` on LHS, align equations

nilsdeppe commented:
@markscheel @carmaza please look at this.

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * \brief A reorientable map from the cube to a frustum.
+ * \image html frustum_table1.png "A frustum with rectangular bases."
+ *
+ * A map from the cube to the volume determined by interpolating between two
```
A map from a cube in rectangular coordinates (\xi, \eta, \zeta) to the volume in rectangular coordinates (x, y, z) determined by interpolating between......

(I'm not familiar with all the documentation in `domain`, so disregard this comment if it's redundant

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * `face_vertices`:
+ *
+ * \f{align*}{
+ * &\textrm{lower x upper base} : x^{+\zeta}_{-\xi,-\eta} = x(-1,-1,1)\\
```
I'm not familiar with what the nomenclature should be here, but I'd probably choose better words to refer to the lower left and the upper right corners of each base. "Left" and "right" are just about as arbitrary as "upper" and "lower" so I'd probably just say "lower left" and "upper right". Up to you.

mar-celine replied:
These variable names do not refer to individual corners of the frustum, but rather to the extents along each dimension, since the bases are rectangular. In other words, the value `lower_x_upper_base` is equal to not only x(-1,-1,1) but also x(-1,1,1) (indeed it is equal to x(-1,eta,1) for all eta, maybe I write that instead?)

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * &\textrm{upper y upper base} : y^{+\zeta}_{+\xi,+\eta} = y(1,1,1)\\
+ * &\textrm{upper y lower base} : y^{-\zeta}_{+\xi,+\eta} = y(1,1,-1)\f}
+ *
+ * As an example, consider a frustum along the z-axis, with the lower base
```
Put this example in a separate note, as it appears to be intended to be an important part what's next in the text, but it's just an example with no posterior use in is paragraph.

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * starting at (x,y) = (-2.0,3.0) and extending to
+ * (2.0,5.0), and with the upper base extending from (0.0,1.0) to (1.0,3.0).
+ * The corresponding value for `face_vertices` is `{{{{-2.0,3.0}}, {{2.0,5.0}},
+ * {{0.0,1.0}}, {{1.0,3.0}}}}`. The user may reorient the frustum by passing
```
Reorienting the frustrum seems like an important option to have. I would move "The user may.... to the constructor." down to a separate paragraph, as it seems to be of no informative use here. Actually, you might want to give a heads-up to the reader by saying something like "Frustrum possesses the following options/members/etc.", then discuss the options.

mar-celine replied:
I like this idea. I am going to come back to this suggestion because I feel it is something that can be added to the documentation for the other maps as well :+1:

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * (2.0,5.0), and with the upper base extending from (0.0,1.0) to (1.0,3.0).
+ * The corresponding value for `face_vertices` is `{{{{-2.0,3.0}}, {{2.0,5.0}},
+ * {{0.0,1.0}}, {{1.0,3.0}}}}`. The user may reorient the frustum by passing
+ * an `OrientationMap` to the constructor. If `with_equiangular_map` is true,
```
link to `OrientationMap` class maybe?

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * The corresponding value for `face_vertices` is `{{{{-2.0,3.0}}, {{2.0,5.0}},
+ * {{0.0,1.0}}, {{1.0,3.0}}}}`. The user may reorient the frustum by passing
+ * an `OrientationMap` to the constructor. If `with_equiangular_map` is true,
+ * then this coordinate map applies a tangent function mapping to the logical
```
This is the first time you mention "logical" in this doc, with no prior definition. Since this class is mostly intended for a map from logical coords, and assuming that the reader does know what logical coords are, I change my first comment above by suggesting "A map from a cube in _logical_ coordinates (\xi, \eta, \zeta)..."

mar-celine replied:
Yes, I agree this is the way to go :smile:

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * \Sigma^z + \Delta^z\zeta\\
+ * \end{bmatrix}\f]
+ *
+ * With Jacobian: \f[J =
```
with

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * {{0.0,1.0}}, {{1.0,3.0}}}}`. The user may reorient the frustum by passing
+ * an `OrientationMap` to the constructor. If `with_equiangular_map` is true,
+ * then this coordinate map applies a tangent function mapping to the logical
+ * \f$\xi\f$ and \f$\eta\f$ coordinates. We refer to the generalized logical
```
We then refer to generalized logical coordinates \f$\Xi\f$ and \f$\mathrm{H}\f$, defined in either case as:

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ *
+ *  \f[\textrm{equiangular eta}  : \mathrm{H}(\eta) = \textrm{tan}(\eta\pi/4)\f]
+ *
+ *  With derivatives:
```
with derivatives

Not very important, I guess, but most of these equations should go with a comma in the end, as equations are usually part of the text.

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ *
+ *  \f[\mathrm{H}'(\eta) = \frac{\pi}{4}(1+\mathrm{H}^2)\f]
+ *
+ *  The equidistant coordinates are:
```
make explicit that "equidistant coordinates" refers to `with_equiangular_map` being `false`, or swap the order in which you introduce each definition of the generalized logical coords.

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ *
+ * In the case where the two
+ * rectangles are geometrically similar, the volume formed is a geometric
+ * frustum. However, this CoordinateMap generalizes to rectangles which need
```
coordinate map, unless a hyperlink is inteded here (which, I think, it's not needed)

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * In the case where the two
+ * rectangles are geometrically similar, the volume formed is a geometric
+ * frustum. However, this CoordinateMap generalizes to rectangles which need
+ * not be similar. In terms of the `face_vertices` variables, we define the
```
I'd introduce the full map first, then say "where we introduced the following auxiliary variables in terms of the `face_vertices` variables", then present the equations for \Sigma^x etc.

mar-celine replied:
In the past I would do this, but I've gotten comments saying that I avoid using variables that aren't defined, even in cases where they're defined immediately after. In this particular case I am also against making this change because I like having the Jacobian presented immediately after the map.

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * \f]
+ *
+ * ### Changing the distribution of gridpoints
+ * By default, \f$\zeta\f$ is used to linearly interpolate between the two
```
By default, the logical corodinate \f$\zeta\f$ is used...

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * normal to the base, and elements near the larger base to be shorter in the
+ * direction normal to the base. We desire elements that have roughly equal
+ * sizes along each of their dimensions, and we obtain this through a
+ * projective map that redistributes the gridpoints in the \f$\zeta\f$
```
I don't know how recurrent "projective map" is in the `domain` documentation, but I'd probably use italics to introduce a new concept that you will further mention in what follows.

mar-celine replied:
This is the first instance of a projective map being used in `domain`, so I should take care to introduce this properly :+1:

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ *
+ * \image html BilinearVProjective.png "Comparison of mappings. (Noah Veltman)"
+ *
+ * As seen Veltman's [Warp-Off]
```
As seen in ?

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * As seen Veltman's [Warp-Off]
+ * (https://bl.ocks.org/veltman/8f5a157276b1dc18ce2fba1bc06dfb48),
+ * we can redistribute the gridpoints using a projective map.
+ * We can also see that in this particular example both maps can leave the
```
We can also see in the figure above (BIlinearVProjective)

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ *
+ * As seen above, \f$w_{\delta}\f$ controls the degree of projection of the
+ * \f$z_2\f$-base of the frustum (seen in purple) into a fourth spatial
+ * dimension \f$w\f$. The resulting grid (seen in red) is obtained by projecting
```
I believe there should be a further description for the relation between this figure and a frustum before diving into these details. Could the purple line be not just the upper base but any intermediate z-constant surface in the frustum? (with its corresponding w, of course.)

mar-celine replied:
I should explain that this diagram is projected downward two spatial dimensions - the entire zeta = +1 plane (the upper base) of the frustum is located at a single point in this diagram. The plane of projection (represented in the diagram by a line) is the entire (x,y,z) volume.

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * affine one. The user can choose to change the `projective_scale_factor`,
+ * increasing the foreshortening in the grid.
+ *
+ * The gridpoint distribution is determined by using \f$\zeta\f$ to linearly
```
This math would be much clearer if there were a figure explaining both x and x tilde, possibly replacing ProjectionOntoPlane.png. It only makes sense to me because you drew the full figure on our whiteboard :-)

mar-celine replied:
I do want to go into more detail explaining x tilde, but I feel that would go beyond the scope of the Frustum map (especially since this projection technique can be used in maps other than Frustum), and I don't know where I would add such documentation. Possibly an additional to the developer guide :smile:

carmaza commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,238 @@ class er;
+ * \end{bmatrix}
+ * \f]
+ *
+ * ### Convenient values
```
Convenient values of w_delta (and possible replacing convenient -> Values of interest)

mar-celine commented:
@carmaza I've created issue #1415 to address your request to provide more detailed documentation on projective mapping, which I feel is outside the scope of the Frustum map. I want to follow your suggestion in making individual map documentation more about the code parameters and member variables a user needs to know to create the correct map, rather than about the math required to derive the map.

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,246 @@ class er;
+ *
+ * \image html ProjectionOntoPlane.png "Controlling the degree of projection."
+ *
+ * As seen above, \f$w_{\delta}\f$ controls the degree of projection of the
```
Need more of a transition. And the z2 base of the frustum is a point, not the purple line, right? Please replace the beginning of this paragraph with "The projection map works by rotating the 3-dimensional frustum into a fourth spatial dimension \f$w\f$.  The figure above is a two-dimensional illustration of the projection map.  In the figure, the frustum is the red line, and the rotated frustum is the purple line.  The parameter \f$w_{\delta}\f$ controls how far the \f$z_2\f$-base of the frustum (a point in the figure) is moved in the \f$w\f$ direction. By default, ..."

(But please correct me if I am still misunderstanding something)

mar-celine replied:
@markscheel I agree this section has a lot that can be misunderstood and improved upon - I think it would be easier to understand the projection method used to obtain the nonuniform grid if instead I talked about rotating a line into a second spatial dimension, because the concepts and equations obtained are the same. But I don't want to clutter the frustum documentation so much with material that doesn't pertain immediately to the map parameters that a user might be interested in though, so I am considering factoring out a lot of this background geometry stuff into a different documentation page. That way the frustum doc can merely refer to it and future maps using this technique can also simply refer to it without restating everything. I've created an issue for this, #1415 so hopefully I can do a better job of explainin the projection stuff in a page dedicated to just that :smile: Thank you for your helpful comments!

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,246 @@ class er;
+ * affine one. The user can choose to change the `projective_scale_factor`,
+ * increasing the foreshortening in the grid.
+ *
+ * The gridpoint distribution is determined by using \f$\zeta\f$ to linearly
```
There is some repetition between here and below without transitions to guide the reader.  

Please change this to "The gridpoint distribution is determined by first defining homogeneous coordinates \f$(\tilde{x},\tilde{y},\tilde{z},w)\f$ that extend from \f$(\vec{x}(\xi,\eta,-1),1)\f$ at \f$\zeta=-1\f$ to \f$(w_{\delta}\vec{x}(\xi,\eta,+1),w_{\delta})\f$ at \f$\zeta=+1\f$ and vary linearly in \f$\zeta\f$ between those endpoints; these coordinates cover the purple line in the figure.  In other words, the homogeneous coordinates are given by <----equation for 4d fustrum in homogeneous coordinates goes here---->."

(and please remove the "...and recovering the Cartesian coordinates..." part at the end of the paragraph.)

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,246 @@ class er;
+ * \frac{1+\zeta}{2}\begin{bmatrix}x_2w_{\delta}\\y_2w_{\delta}\\
+ * z_2w_{\delta}\\w_{\delta}\\\end{bmatrix}\f]
+ *
+ * The equation for the projected 3D frustum in Cartesian coordinates is
```
"3D frustum" -> "3D frustum (red line)"

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,246 @@ class er;
+ * \end{bmatrix}\f]
+ *
+ * ### Generalized frustum map
+ * We define an auxiliary variable `projective_zeta` \f$ = \mathrm{Z}\f$ such
```
Please add a new sentence at the beginning "Now we define a generalized Frustum that includes a projective map."

And please remove the stuff about the Cartesian coordinates linear in Z and the stuff about Z being rationally interpolated, which doesn't make sense until you see the equation below and just trips up the reader.  I suggest "We define an auxiliary variable `projective_zeta` \f$ = \mathrm{Z}\f$ by <----equation goes here---->, where \f$w_{\pm} := w_{\delta}\pm 1\f$.  The derivative of \f$\mathrm{Z}\f$ is <----equation goes here---->".

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,246 @@ class er;
+ * \end{bmatrix}
+ * \f]
+ *
+ * ### Values of interest
```
"Values of Interest" -> "Suggested values for \f$w_{\delta}\f$ in the projective map"

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,246 @@ class er;
+ * ### Values of interest
+ * \image html ProjectionOfCube.png "Trapezoids from squares. (Davide Cervone)"
+ *
+ * As seen in Cervone's [Cubes and Hypercubes Rotating]
```
Add this at the beginning so the reader knows what is going on: "When constructing a Frustum map, it is not immediately obvious what value of \f$w_{\delta}\f$ to use in the projective map; here we present a choice of \f$w_{\delta}\f$ that will produce minimal grid distortion.  We cover two cases: the first is the special case in which the two rectangular Frustum bases are related by a simple scale factor; the second is the general case."

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,246 @@ class er;
+ *
+ * where \f$\frac{L_1}{L_2}\f$ is the ratio between any pair of
+ * corresponding side lengths of the \f$z_1\f$ and \f$z_2\f$-bases of the
+ * frustum, respectively. Other values for `projective_scale_factor` are
```
I don't understand the sentence "Other values for `projective_scale_factor` are...".  There is a general formula below. I suggest omitting this sentence.

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,246 @@ class er;
+ * are not related by a scaling, for which there is not a unique
+ * `projective_scale_factor` value.
+ *
+ * For the general case one will want to use the value:
```
This is a complicated enough formula that it doesn't make sense to make a user compute it and possibly get it wrong.  Can you add a constructor to Frustum that simply uses this value?  Or maybe even make it the default value?

markscheel reviewed (changes requested):
This is very much improved, especially the nice figures! I also like how you explain the map first without the projective map and then add the projective map later. I do have a few more clarity improvements.

mar-celine commented:
@markscheel @carmaza I've rebased this on develop after the merging of #1421. There is a new non-fixup commit that adds a documentation page to the developer guide, "Add Redistributing Gridpoints to DevGuide", please look at it! Also please let me know if I can squash in the fixups/if there are any fixups you don't want squashed in yet.

carmaza commented:
@gsb76 you can squash the fixup addressing my comments.

markscheel commented on src/Domain/CoordinateMaps/Frustum.hpp:
```diff
@@ -21,27 +21,155 @@ class er;
+ * `projective_scale_factor` is set to a quantity other than unity, then this
+ * coordinate map applies a rational function mapping to the logical \f$\zeta\f$
+ * coordinate. This generalized logical coordinate is referred to as
+ * \f$mathrm{Z}\f$. If `use_suggested_value` is `true`, the user-specified
```
Backslash in front of mathrm.

markscheel commented on src/Domain/CoordinateMaps/Frustum.cpp:
```diff
@@ -22,10 +22,13 @@
-                 const bool with_equiangular_map) noexcept
+                 const bool with_equiangular_map,
+                 const double projective_scale_factor,
+                 const bool use_suggested_value) noexcept
```
Could use better option name that answers the question "suggested value of what"?  How about `auto_projective_scale_factor` or `use_auto_projective_scale_factor`?

markscheel reviewed (commented):
A few more minor comments.  I am ok with squashing.

Is it easy to add a test that tests the option that gives the suggested value of projective_scale_factor?

mar-celine committed:
Add Redistributing Gridpoints to DevGuide

mar-celine committed:
Add documentation and projective_scale_factor to Frustum

mar-celine committed:
Add FrustalCloak domain creator

mar-celine commented:
@markscheel I've added a test in Test_Frustum.cpp:
```
void test_auto_projective_scale_factor() {
  INFO("Auto projective scale factor");
  // This test tests that the correct suggested projective scale factor
  // is computed based on the side lengths of the frustum.

  // Set up random number generator
  MAKE_GENERATOR(gen);
  std::uniform_real_distribution<> lower_bound_dis(-14, -2);
  std::uniform_real_distribution<> upper_bound_dis(2, 14);
  std::uniform_real_distribution<> logical_dis(-1, 1);

  const double lower_x_lower_base = lower_bound_dis(gen);
  CAPTURE_PRECISE(lower_x_lower_base);
  const double lower_y_lower_base = lower_bound_dis(gen);
  CAPTURE_PRECISE(lower_y_lower_base);
  const double upper_x_lower_base = upper_bound_dis(gen);
  CAPTURE_PRECISE(upper_x_lower_base);
  const double upper_y_lower_base = upper_bound_dis(gen);
  CAPTURE_PRECISE(upper_y_lower_base);
  const double lower_x_upper_base = lower_bound_dis(gen);
  CAPTURE_PRECISE(lower_x_upper_base);
  const double lower_y_upper_base = lower_bound_dis(gen);
  CAPTURE_PRECISE(lower_y_upper_base);
  const double upper_x_upper_base = upper_bound_dis(gen);
  CAPTURE_PRECISE(upper_x_upper_base);
  const double upper_y_upper_base = upper_bound_dis(gen);
  CAPTURE_PRECISE(upper_y_upper_base);
  const double xi = logical_dis(gen);
  CAPTURE_PRECISE(xi);
  const double eta = logical_dis(gen);
  CAPTURE_PRECISE(eta);
  const double zeta = logical_dis(gen);
  CAPTURE_PRECISE(zeta);

  const std::array<double, 3> logical_coord{{0.0, 0.0, zeta}};
  const double lower_bound = 2.0;
  const double upper_bound = 5.0;
  const double sigma_z = 0.5 * (upper_bound + lower_bound);
  const double delta_z = 0.5 * (upper_bound - lower_bound);
  const double w_delta = sqrt(((upper_x_lower_base - lower_x_lower_base) *
                               (upper_y_lower_base - lower_y_lower_base)) /
                              ((upper_x_upper_base - lower_x_upper_base) *
                               (upper_y_upper_base - lower_y_upper_base)));
  const double projective_zeta = (w_delta - 1.0 + zeta * (w_delta + 1.0)) /
                                 (w_delta + 1.0 + zeta * (w_delta - 1.0));
  const double expected_physical_z = sigma_z + delta_z * (projective_zeta);
  const std::array<std::array<double, 2>, 4> face_vertices{
      {{{lower_x_lower_base, lower_y_lower_base}},
       {{upper_x_lower_base, upper_y_lower_base}},
       {{lower_x_upper_base, lower_y_upper_base}},
       {{upper_x_upper_base, upper_y_upper_base}}}};

  const CoordinateMaps::Frustum map(face_vertices, lower_bound, upper_bound, {},
                                    false, 1.0, true);  // Upper Z frustum
  CHECK(map(logical_coord)[2] == approx(expected_physical_z));
}
```

I'm putting a copy of it here because I've squashed in the change into the commit that adds the projective map. This separately computes the value for `projective_scale_factor` and checks that it is the same as the one computed by the map by checking that it maps the zeta coordinates in the same way. Hope this is good :+1:  I've addressed your other comments as well and squashed them in, and I haven't made any further changes/additions since the last time you looked. Hopefully this can be merged soon!

kidder commented:
@markscheel please look at

markscheel reviewed (approved):
Yay!
